File: solutions/fsharp/kindergarten-garden/KindergartenGarden.fs
module KindergartenGarden

type Plant =
    | Grass
    | Clover
    | Radishes
    | Violets

let private toPlant =
    function
    | 'G' -> Grass
    | 'C' -> Clover
    | 'R' -> Radishes
    | 'V' -> Violets
    | symbol -> failwith $"Unknown symbol '{symbol}'"
    
let plants (diagram: string) (student:string) =
    let startIndex = 2 * int (student[0] - 'A')
    
    diagram.Split('\n')
    |> Seq.collect _.Substring(startIndex, 2)
    |> Seq.map toPlant
    |> Seq.toList

File: solutions/python/kindergarten-garden/kindergarten_garden.py
PLANT_CODES = {
    "C": "Clover",
    "G": "Grass",
    "R": "Radishes",
    "V": "Violets"
}

DEFAULT_STUDENTS = (
    "Alice", "Bob", "Charlie", "David", "Eve", "Fred",
    "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"
)


class Garden:
    def __init__(self, diagram: str, students: tuple[str, ...] = DEFAULT_STUDENTS):
        self.rows = diagram.splitlines()
        self.students = sorted(students)

    def plants(self, student: str) -> list[str]:
        index = self.students.index(student)
        start = index * 2
        end = start + 2
        row1_segment = self.rows[0][start:end]
        row2_segment = self.rows[1][start:end]
        return [PLANT_CODES[code] for code in row1_segment + row2_segment]
